File: java21/src/main/java/io/github/sps23/designpatterns/adapter/LegacyBankPaymentAdapter.java
package io.github.sps23.designpatterns.adapter;

import java.util.Locale;

sealed interface PaymentGateway permits LegacyBankPaymentAdapter {
    /**
     * Charges a customer using the normalized application-level payment model.
     */
    PaymentResult charge(PaymentRequest request);
}

/**
 * Business-level payment request understood by checkout and other application
 * services.
 */
record PaymentRequest(String customerId, int amountInCents, String currency) {
}

/**
 * Business-level payment result returned to the application after legacy
 * translation.
 */
record PaymentResult(boolean approved, String transactionId, String message) {
}

/**
 * Legacy bank response format (status code + textual details).
 */
record LegacyBankResponse(String statusCode, String reference, String detail) {
}

/**
 * Simulates a legacy banking API with protocol-specific parameter names and
 * status codes.
 */
final class LegacyBankApi {
    /**
     * Legacy contract where "00" means approved and other codes represent
     * rejections.
     */
    LegacyBankResponse submitPayment(String clientCode, long minorUnits, String isoCurrency) {
        if (minorUnits <= 0) {
            return new LegacyBankResponse("12", "N/A", "Amount must be greater than zero");
        }
        if (!"USD".equals(isoCurrency) && !"EUR".equals(isoCurrency)
                && !"GBP".equals(isoCurrency)) {
            return new LegacyBankResponse("14", "N/A", "Unsupported currency: " + isoCurrency);
        }
        return new LegacyBankResponse("00",
                "TX-" + clientCode.toUpperCase(Locale.ROOT) + "-" + minorUnits, "Approved");
    }
}

/**
 * Adapter pattern in Java 21.
 *
 * Wraps a legacy banking API and exposes a modern payment-gateway interface for
 * checkout services.
 */
public final class LegacyBankPaymentAdapter implements PaymentGateway {
    private final LegacyBankApi legacyBankApi;

    public LegacyBankPaymentAdapter(LegacyBankApi legacyBankApi) {
        this.legacyBankApi = legacyBankApi;
    }

    @Override
    public PaymentResult charge(PaymentRequest request) {
        if (request == null) {
            throw new IllegalArgumentException("Payment request must not be null");
        }
        if (request.customerId() == null || request.customerId().isBlank()) {
            throw new IllegalArgumentException("Customer id must not be blank");
        }
        if (request.amountInCents() <= 0) {
            throw new IllegalArgumentException("Amount must be greater than zero");
        }

        String currency = normalizeCurrency(request.currency());
        LegacyBankResponse response = legacyBankApi.submitPayment(request.customerId(),
                request.amountInCents(), currency);

        boolean approved = "00".equals(response.statusCode());
        return new PaymentResult(approved, response.reference(),
                approved ? "Payment approved" : response.detail());
    }

    private String normalizeCurrency(String currency) {
        if (currency == null || currency.isBlank()) {
            throw new IllegalArgumentException("Currency must not be blank");
        }
        return currency.trim().toUpperCase(Locale.ROOT);
    }
}

File: kotlin/src/main/kotlin/io/github/sps23/designpatterns/adapter/LegacyBankPaymentAdapter.kt
package io.github.sps23.designpatterns.adapter

import java.util.Locale

sealed interface PaymentGateway {
    fun charge(request: PaymentRequest): PaymentResult
}

data class PaymentRequest(
    val customerId: String,
    val amountInCents: Int,
    val currency: String,
)

data class PaymentResult(
    val approved: Boolean,
    val transactionId: String,
    val message: String,
)

data class LegacyBankResponse(
    val statusCode: String,
    val reference: String,
    val detail: String,
)

class LegacyBankApi {
    fun submitPayment(
        clientCode: String,
        minorUnits: Long,
        isoCurrency: String,
    ): LegacyBankResponse =
        when {
            minorUnits <= 0 -> LegacyBankResponse("12", "N/A", "Amount must be greater than zero")
            isoCurrency !in setOf("USD", "EUR", "GBP") ->
                LegacyBankResponse("14", "N/A", "Unsupported currency: $isoCurrency")
            else -> LegacyBankResponse("00", "TX-${clientCode.uppercase()}-$minorUnits", "Approved")
        }
}

/**
 * Adapter pattern in Kotlin.
 *
 * Wraps a legacy banking API and exposes a modern payment-gateway interface for
 * checkout services.
 */
class LegacyBankPaymentAdapter(
    private val legacyBankApi: LegacyBankApi,
) : PaymentGateway {
    override fun charge(request: PaymentRequest): PaymentResult {
        require(request.customerId.isNotBlank()) { "Customer id must not be blank" }
        require(request.amountInCents > 0) { "Amount must be greater than zero" }

        val currency = normalizeCurrency(request.currency)
        val response =
            legacyBankApi.submitPayment(
                clientCode = request.customerId.trim(),
                minorUnits = request.amountInCents.toLong(),
                isoCurrency = currency,
            )

        val approved = response.statusCode == "00"
        return PaymentResult(
            approved = approved,
            transactionId = response.reference,
            message = if (approved) "Payment approved" else response.detail,
        )
    }

    private fun normalizeCurrency(currency: String): String {
        require(currency.isNotBlank()) { "Currency must not be blank" }
        return currency.trim().uppercase(Locale.ROOT)
    }
}
